lib/internal/main/run.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+/* eslint-disable node-core/prefer-primordials */
+
+// There is no need to add primordials to this file.
+// `run.js` is a script only executed when `node run <script>` is called.
+const {
+  prepareMainThreadExecution,
+  markBootstrapComplete,
+} = require('internal/process/pre_execution');
+const { getPackageJSONScripts } = internalBinding('modules');
```

Comment:
Can't the be done in JS?

Reply:
Yes, it can, but also the whole file can be written in C++ as well.

Reply:
I think we should ask ourselves a different question: "is there a benefit to implementing this in JS?" which changes our mindset to have a default implementation in C++, where some features come out of box - such as not thinking about primordials, performance etc.

To answer your question, there is no significant reason other than shaving couple of `ms`

Reply:
>I think we should ask ourselves a different question: "is there a benefit to implementing this in JS?" which changes our mindset to have a default implementation in C++,

I strongly disagree with this mindset. It alienates 90% of our contributors.

Reply:
As far as making it fast goes, I don't think parsing the JSON in C++ makes that big of a difference. My [original pull request](https://github.com/nodejs/node/pull/46534) just did good old `JSON.parse()` and it was 35ms compared to 330ms from `npm run`. I've analyzed in https://github.com/nodejs/node/pull/52190#issuecomment-2016528727 why this and the other pull request are faster - because we are doing less than `npm run` by not loading npm itself, and cutting off some features that npm supports.

Reply:
Although I'll add, if we want to reach the speed of e.g. deno run, then it needs to be implemented in C++ entirely. Though as I said in https://github.com/nodejs/node/pull/46534#issue-1573511033 I am not sure if 35ms -> 5ms is really worth rewriting them all in C++...while 330ms -> 35ms with just a <100 lines of JS in Node.js looks more like a worthy thing.

Reply:
In order to reduce 30ms to 5ms we need to move the function to C++. Matteo and several others recommending moving all to C++. I agree, and I'll follow up on a different pull-request. For the time being, I'll keep as it is since this is not a blocking comment.